src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,853 @@
+        if (r < 0)
+               return r;
+
+        if (add) {
+                 r = sd_nfnl_nft_message_new_setelems_begin(nfnl, &transaction[1], nft_systemd_table_name, "masq_saddr");
+                 if (r < 0)
+                        return r;
+        } else {
+                 r = sd_nfnl_nft_message_del_setelems_begin(nfnl, &transaction[1], nft_systemd_table_name, "masq_saddr");
+                 if (r < 0)
+                        return r;
+        }
+
+        r = nft_message_add_setelem_iprange(transaction[1], source, source_prefixlen);
+        if (r < 0)
+               return r;
+
+        r = sd_nfnl_message_batch_end(nfnl, &transaction[2]);
+        if (r < 0)
+               return r;
+
+        r = nfnl_netlink_sendv(nfnl, transaction, 3);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+int fw_nftables_add_local_dnat(
+                bool add,
+                int af,
+                int protocol,
+                uint16_t local_port,
+                const union in_addr_union *remote,
+                uint16_t remote_port,
+                const union in_addr_union *previous_remote) {
+        uint32_t data[2], key[2];
+        sd_netlink_message *transaction[4];
+        unsigned int tsize;
+        int r;
+
+        assert(add || !previous_remote);
+
+        if (af != AF_INET)
+                return -EOPNOTSUPP;
+
+        if (!IN_SET(protocol, IPPROTO_TCP, IPPROTO_UDP))
+                return -EOPNOTSUPP;
+
+        if (local_port <= 0)
+                return -EINVAL;
+
+        key[0] = protocol;
+        key[1] = htobe16(local_port);
+
+        if (!remote)
+                return -EOPNOTSUPP;
+
+        if (remote_port <= 0)
+                return -EINVAL;
+
+        data[1] = htobe16(remote_port);
+
+        memset(transaction, 0, sizeof(transaction));
+        r = sd_nfnl_message_batch_begin(nfnl, &transaction[0]);
+        if (r < 0)
+                return r;
+
+        tsize = 1;
+        /* If a previous remote is set, remove its entry */
+        if (add && previous_remote && previous_remote->in.s_addr != remote->in.s_addr) {
+                data[0] = previous_remote->in.s_addr;
+
+                nft_del_element(&transaction[tsize], nft_systemd_dnat_map_name, key, sizeof(key), data, sizeof(data));
+                tsize++;
+        }
+
+        data[0] = remote->in.s_addr;
+
+        if (add)
+                nft_add_element(&transaction[tsize], nft_systemd_dnat_map_name, key, sizeof(key), data, sizeof(data));
+        else
+                nft_del_element(&transaction[tsize], nft_systemd_dnat_map_name, key, sizeof(key), data, sizeof(data));
+
+        tsize++;
+
+        r = sd_nfnl_message_batch_end(nfnl, &transaction[tsize]);
+        if (r < 0)
+                return r;
+
+        tsize++;
+        r = nfnl_netlink_sendv(nfnl, transaction, tsize);
+        if (r < 0)
+                return r;
+
+        return 0;
```

Comment:
the messages are leaked?

Reply:
I will run the existing test case through valgrind to make sure this is resolved correctly.

Reply:
Indeed, I've fixed this. valgrind --leak-check=full ./build/test-firewall-util
now reports "All heap blocks were freed".